specs/merkle-tree.md

```diff
@@ -0,0 +1,166 @@
+
+
+## High-Level Merkle Tree Invariants
+The purpose of these invariants is to define an unambiguous, consistent structure for Sphinx Merkle trees. This ensures that it's easy for off-chain tooling to recreate the exact Merkle tree signed by the user, and to check that a given Merkle tree is valid. It's worth mentioning that it's possible to construct Merkle trees that do not adhere to this spec, but are executable on-chain. We discuss this possibility and the expected executor behavior in more detail in the [Assumptions section](#assumptions).
+
+### 1. Must be executable on all chains for which there is at least one Merkle leaf.
+We assume that the input Merkle leaf(s) data satisfies the on-chain conditions in the corresponding `SphinxModuleProxy` contract(s) and that the executor is not buggy. See the [Assumptions section](#assumptions) for more information.
+
+### 2. There must be exactly one `APPROVE` leaf or `CANCEL` leaf per chain
+Constructing a tree that contains `APPROVE` or `CANCEL` leaves that approve or cancel multiple separate deployments on a single chain would lead to an ambiguous situation for the executor where it is unclear which leaf should be approved or canceled. So we impose the restriction that in a valid Merkle tree there must be exactly one `APPROVE` or `CANCEL` leaf per chain.
+
+If the `arbitraryChain` field is set to true, then we consider this approval leaf to apply to all chains. Therefore if `arbitraryChain` = true, there should be exactly one `APPROVE` leaf in the entire tree and no `CANCEL` leaves.
+
+### 3. `EXECUTE` leaves must always follow an `APPROVE` leaf on the same chain
+To ensure that every `EXECUTE` leaf has been approved by the `SphinxModuleProxy` owners and can be executed, all `EXECUTE` leaves must follow an `APPROVE` leaf on the same chain which performs this verification.
+
+### 4. `EXECUTE` leaves must *never* exist in a tree with a `CANCEL` leaf on the same chain
+To ensure that all `EXECUTE` leaves are executable, they must *never* follow a `CANCEL` leaf on the same chain. The presence of a `CANCEL` leaf for a given chain [implies there is no `APPROVE` leaf](#2-there-must-be-exactly-one-approve-leaf-or-cancel-leaf-per-chain), and therefore any `EXECUTE` leaves on that chain [will not be executable](#3-execute-leaves-must-always-follow-an-approve-leaf-on-the-same-chain).
+
+### 5. Every leaf `chainId` and `index` combination must be unique within the tree
+Constructing a tree with multiple leaves with the same chain ID and index would make it ambiguous which leaf should be executed, and would result in a tree that cannot be fully executed. So we impose the restriction that in a valid Merkle tree there must be at most one leaf with each chain ID and index combination which ensures there is an unambiguous ordering of leaves within the tree.
+
+### 6. Each leaf on a chain must have an index field with a value that starts with 0 and sequentially increments by 1
```

Comment:
> Each leaf on a chain must have an index field with a value that starts with 0

This sentence is confusing because each leaf on a chain doesn't have an index field that starts with 0. Rather, the `APPROVE` and `CANCEL` leaves have an index that equals zero

Reply:
So since you seem really opinionated on this, I'm fine with splitting it up for clarity. We can say this: 
- EACH `APPROVE` or `CANCEL` leaf on a chain must have an index of 0
- Each sequence of `EXECUTE` leaves on a chain must have the index of the first `EXECUTE` leaf be 1, and the index of the following leaves sequentially increment by 1

However, I just want to point out that the above two invariants is functionally equivalent to the original one you're responding too. A more clear version might make this my point easier to see:
- Each sequence of leaves with the same `chainid` must have the `index` of the first leaf be 0, and the `index` of the following leaves sequentially increment by 1